backend/network.py

```
@@ -0,0 +1,184 @@
+            self._tunnel()
+        self.sock = self._context.wrap_socket(
+            self.sock, server_hostname=self._tls_server_hostname
+        )
+
+
+def _port(parts: urllib.parse.SplitResult) -> int:
+    """Return an explicit valid TCP port, or the scheme default."""
+    try:
+        port = parts.port
+    except ValueError:
+        raise errors.blocked_url("invalid port")
```

Ruff (0.16.1)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)